osier/equations.py

```diff
@@ -0,0 +1,160 @@
+from abc import ABC, abstractmethod
+import numpy as np
+
+class OsierEquation(ABC):
+
+    def __init__(self) -> None:
+        pass        
+
+    @abstractmethod
+    def _do(self, technology_list, X):
+        pass
+
+
+def get_tech_names(technology_list):
+    """
+    Parameters
+    ----------
+    technology_list : list of :class:`osier.Technology` objects
+        The list of technologies.
+    dispatch_results : :class:`pandas.DataFrame`
+        The results dispatch results for each technology in 
+        :attr:`technology_list` from a :class:`osier.DispatchModel` run.
+    """
+
+    tech_names = [t.technology_name for t in technology_list]
+
+    return tech_names
+
+
+def get_dispatchable_techs(technology_list):
+    """
+    Parameters
+    ----------
+    technology_list : list of :class:`osier.Technology` objects
+        The list of technologies.
+    dispatch_results : :class:`pandas.DataFrame`
+        The results dispatch results for each technology in 
+        :attr:`technology_list` from a :class:`osier.DispatchModel` run.
+    """
+
+    dispatchable_techs = [t for t in technology_list if t.dispatchable]
+
+    return dispatchable_techs
+
+
+def get_nondispatchable_techs(technology_list):
+    """
+    Parameters
+    ----------
+    technology_list : list of :class:`osier.Technology` objects
+        The list of technologies.
+    dispatch_results : :class:`pandas.DataFrame`
+        The results dispatch results for each technology in 
+        :attr:`technology_list` from a :class:`osier.DispatchModel` run.
+    """
+
+    dispatchable_techs = [t for t in technology_list if not t.dispatchable]
+
+    return dispatchable_techs    
+
+
+def get_dispatchable_names(technology_list):
+    """
+    Parameters
+    ----------
+    technology_list : list of :class:`osier.Technology` objects
+        The list of technologies.
+    dispatch_results : :class:`pandas.DataFrame`
+        The results dispatch results for each technology in 
+        :attr:`technology_list` from a :class:`osier.DispatchModel` run.
+    """
+
+    dispatchable_techs = [t.technology_name for t in technology_list if t.dispatchable]
+
+    return dispatchable_techs
+
+
+def annualized_capital_cost(technology_list, dispatch_model=None):
+    """
+    Parameters
+    ----------
+    technology_list : list of :class:`osier.Technology` objects
+        The list of technologies.
+    dispatch_model : :class:`osier.DispatchModel`
+        A _solved_ dispatch model (i.e. with model results and objective
+        attributes).
```

Comment:
```suggestion
def annualized_capital_cost(technology_list):
    """
    Parameters
    ----------
    technology_list : list of :class:`osier.Technology` objects
        The list of technologies.
```

Reply:
The reason I have `dispatch_model=None` is due to the way I implemented objectives + constraints. The idea is to allow users to create their own objective or constraint functions, which get passed to the model class. There are only two things needed to calculate any quantity of interest -- the technology list and the dispatch results. However, not all objectives need both -- but I can't know that _a priori_ and it's simpler to just always pass both to the desired function rather than somehow check which pieces of data the function needs. (The metadata could about what info the function needs could be contained in a class like `OsierEquation` but I think that's still more difficult to implement and confusing to use). 

TL;DR The `dispatch_model` param is necessary.

Reply:
It sound like what you're saying is that you anticipate someone will extend this class and write their own `annualized_capital_cost` function that requires using the `dispatch_model` parameter, right?

Reply:
A couple of things 
a) it's not a class -- just a function
b) users can write their own capital cost function, or use the one provided. 
c) In _general_ users can write any function they want that uses either a technology list, dispatch model, or both. These functions are passed as either objective arguments or constraint arguments to the `CapacityExpansion` class.
d) The `CapacityExpansion` class loops through the provided list of objectives and constraints, and assumes that they will accept a `technology_list` and a `dispatch_model`.